design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,150 @@
+# Extensible configuration management
+
+* Date: __2025-08-12__
+* Status: __New__
+
+## Background
+
+The built-in configuration and log management features offered by `tedge-agent` has the following limitations:
+
+- only file based configurations and logs can be retrieved.
+- while updating a configuration file, there are no pre/post actions that can be performed.
+  For e.g, a user would want to backup the existing configuration before updating the configuration,
+  and restart a service after the configuration is updated.
+- Though the tedge workflows feature allows overriding the existing behavior,
+  when a built-in operation is overridden, you end up losing all existing functionality,
+  like uploading/downloading the files from the cloud as well.
+
+The goal is to provide an interface that allows users to extend only the specific aspect of the operation that they want,
+like the pre-update or post-update actions, leaving the rest of the behavior unchanged.
+
+## Solution proposal
+
+- The agent supports plugins for log and config management.
+- The supported types for the respective operation is gathered from all these plugins.
+- The agent maintains a mapping of the types and their corresponding plugins.
+- When an an operation is received for a type, it is delegated to the corresponding plugin.
+
+### Log Management
+
+- `tedge-agent` to add support for multiple log management plugins defined under `/etc/tedge/log-plugins`.
```

Comment:
This path must be discussed as `/etc/tedge` doesn't feel like the right location for binaries/scripts. I just kept it consistent with `sm-plugins` for now.

The advantage with this location is during A/B firmware updates where the contents of `/etc/tedge` are always backed up and restored on the new partition preserving any custom plugins as well. The downside with this is that an updated built-in plugin on the new partition might get replaced by the old one.

Reply:
I totally agree, the plugin location is a matter of discussion.

I would stop using `/etc/tedge` to store executables plugins for the new features. `/usr/lib` is probably the right location. Even I prefer moving `/etc/tedge/sm-plugins` to `/usr/lib/tedge/sm-plugins`.
For `sm-plugins`, to keep backwards compatibility, we can read multiple locations with priority.

Reply:
The whole concept of layered configuration was also proposed by a user on Discord. https://uapi-group.org/specifications/specs/configuration_files_specification/

Reply:
Personally, I don't like deep file hierarchies and would remove the `agent` sub-directory.

Reply:
Resolved by 1d7600b55